File: MQTT/MQTT.py
# Relay node that subscribes to the broker
# Subscribe to HiveMQ broker and grab the potentiometer data. Then, push this
# data to a Hostinger database.

import paho.mqtt.client as mqtt
import mysql.connector
from datetime import datetime

# HiveMQ settings
MQTT_SERVER = "broker.mqtt-dashboard.com"
MQTT_PORT = 1883
MQTT_TOPIC = "testtopic/temp/outTopic/potvalue"

# Hostinger database settings
DB_HOST = ""  # Replace with your database host
DB_USER = ""       # Replace with your database username
DB_PASSWORD = ""   # Replace with your database password
DB_NAME = ""           # Replace with your database name

# Callback when connected
def on_connect(client, userdata, flags, rc):
    if rc == 0:
        print("Connected to MQTT broker")
        client.subscribe(MQTT_TOPIC)
    else:
        print(f"Failed to connect, return code {rc}")

# Callback when message is received
def on_message(client, userdata, message):
    try:
        payload = str(message.payload.decode("utf-8"))
        print(f"Received data: {payload}")

        # Process only messages containing potentiometer data
        if "Pot Value:" in payload:
            pot_value = payload.split(":")[1].strip()  # Extract the numeric value
            print(f"Extracted Potentiometer Value: {pot_value}")

            # Insert data into Hostinger database
            try:
                conn = mysql.connector.connect(
                    host=DB_HOST,
                    user=DB_USER,
                    password=DB_PASSWORD,
                    database=DB_NAME
                )
                cursor = conn.cursor()
                
                # Define your table and fields; ensure the table exists
                insert_query = "INSERT INTO potentiometer_data (value, timestamp) VALUES (%s, %s)"
                timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
                cursor.execute(insert_query, (pot_value, timestamp))
                conn.commit()
                print("Data inserted into Hostinger database successfully.")
            except mysql.connector.Error as db_err:
                print(f"Database error: {db_err}")
            finally:
                if conn.is_connected():
                    cursor.close()
                    conn.close()
        else:
            print("Message does not contain potentiometer data. Ignoring.")
    except Exception as e:
        print(f"Error processing message: {e}")

# MQTT Client setup
client = mqtt.Client()
client.on_connect = on_connect
client.on_message = on_message

# Connect to HiveMQ broker
client.connect(MQTT_SERVER, MQTT_PORT, 60)
client.loop_forever()

File: MQTT/mqtt.php
<?php
// Enable error reporting
ini_set('display_errors', 1);
ini_set('display_startup_errors', 1);
error_reporting(E_ALL);


// Database connection details
$host = "localhost"; // Replace with your database host
$username = ""; // Replace with your database username
$password = ''; // Replace with your database password
$database = ""; // Replace with your database name

// Create connection
$conn = new mysqli($host, $username, $password, $database);

// Check connection
if ($conn->connect_error) {
    echo "Connection error.";
    die("Connection failed: " . $conn->connect_error);
}

$potValue = NULL;

foreach ($_REQUEST as $key => $value) {
    if ($key == "pot") {
        $potValue = $value;
    }
}

if (isset($potValue)) {
    $sql = "INSERT INTO `potentiometer_data`(`value`) VALUES ('" . $potValue . "')";
    if (!$conn->query($sql)) {
        echo "Error with SQL: " . $conn->error;
    }
}

// Query to get the newest value from the table
$sql = "SELECT `value` FROM `potentiometer_data` ORDER BY `id` DESC LIMIT 1";
$result = $conn->query($sql);

if ($result && $result->num_rows > 0) {
    $row = $result->fetch_assoc();
    echo "<p>Newest Potentiometer Value: " . htmlspecialchars($row['value']) . "</p>";
} else {
    echo "<p>No data available or query failed.</p>";
}

$conn->close();
?>

</body>
</html>

